qiita_client/plugin.py

```diff
@@ -0,0 +1,330 @@
+        The default parameter sets of the command, keyed by parameter set name.
+        The values should be a dictionary in which keys are the parameter names
+        and values are the specific value for each parameter
+
+    Raises
+    ------
+    TypeError
+        If `function` is not callable
+    ValueError
+        If `function` does not accept 4 parameters
+    """
+    def __init__(self, name, description, function, required_parameters,
+                 optional_parameters, outputs, default_parameter_sets=None):
+        self.name = name
+        self.description = description
+
+        # Make sure that `function` is callable
+        if not callable(function):
+            raise TypeError(
+                "Couldn't create command '%s': the provided function is not "
+                "callable (type: %s)" % (name, type(function)))
+
+        # `function` will be called with the following Parameters
+        # qclient, job_id, job_parameters, output_dir
+        # Make sure that `function` can receive 4 parameters
+        if function.__code__.co_argcount != 4:
+            raise ValueError(
+                "Couldn't register command '%s': the provided function does "
+                "not accept 4 parameters (number of parameters: %d)"
+                % (name, function.__code__.co_argcount))
+
+        self.function = function
+        self.required_parameters = required_parameters
+        self.optional_parameters = optional_parameters
+        self.default_parameter_sets = default_parameter_sets
+        self.outputs = outputs
+
+    def __call__(self, qclient, server_url, job_id, output_dir):
+        return self.function(qclient, server_url, job_id, output_dir)
+
+
+class QiitaArtifactType(object):
+    """A Qiita artifact type
+
+        Parameters
+        ----------
+        name : str
+            The artifact type name
+        description : str
+            The artifact type description
+        can_be_submitted_to_ebi : bool
+            Whether the artifact type can be submitted to EBI or not
+        can_be_submitted_to_vamps : bool
+            Whether the artifact type can be submitted to VAMPS or not
+        filepath_types : list of (str, bool)
+            The list filepath types that the new artifact type supports, and
+            if they're required or not in an artifact instance of this type"""
+    def __init__(self, name, description, can_be_submitted_to_ebi,
+                 can_be_submitted_to_vamps, filepath_types):
+        self.name = name
+        self.description = description
+        self.ebi = can_be_submitted_to_ebi
+        self.vamps = can_be_submitted_to_vamps
+        self.fp_types = filepath_types
+
+
+class BaseQiitaPlugin(object):
+    def __init__(self, name, version, description, publications=None):
+        self.name = name
+        self.version = version
+        self.description = description
+        self.publications = dumps(publications) if publications else ""
```

Comment:
There is going to be cases we are going to be using tools without publication, the first one that comes to mind is: KneedData.